In [54]:
import pandas as pd 
import seaborn as sns
import numpy as np

In [2]:
df = pd.read_csv('LoanData.csv', low_memory=False)
df.head()

ReportAsOfEOD                                LoanId  LoanNumber  \
0    2022-06-11  66AE108B-532B-4BB3-BAB7-0019A46412C1      483449   
1    2022-06-11  A6635EA6-2F39-4DEA-AA3B-001C9521BE7C     2819530   
2    2022-06-11  D152382E-A50D-46ED-8FF2-0053E0C86A70      378148   
3    2022-06-11  87342E13-66CB-483F-833A-007953E50C78      451831   
4    2022-06-11  87227056-6BF9-410C-98D1-008F788E122A      349381   

           ListedOnUTC     BiddingStartedOn  BidsPortfolioManager  BidsApi  \
0  2016-03-23 16:07:19  2016-03-23 16:07:19                   970   1150.0   
1  2022-05-31 15:11:34  2022-05-31 18:11:34                    16      0.0   
2  2015-06-25 11:02:28  2015-06-25 11:02:28                  1295      0.0   
3  2016-01-14 10:00:21  2016-01-14 10:00:21                  2700    565.0   
4  2015-03-24 15:55:44  2015-03-24 15:55:44                  1115      0.0   

   BidsManual                                 PartyId  NewCreditCustomer  ...  \
0         5.0  {EBF05573-554D-4A3B-BC77-A2CF00B7D110}              False  ...   
1         0.0  {F6507F91-F992-41FA-BB4C-AC37009B3E58}              False  ...   
2      1705.0  {46C6CBA4-0FBE-44AD-9304-A3EF0111A5FB}              False  ...   
3      5835.0  {CA64DA9B-8E95-450E-9EFE-A58601016DB2}               True  ...   
4       385.0  {F08F654D-DB2E-4C4B-8C90-A46100FCE7B6}               True  ...   

  PreviousEarlyRepaymentsCountBeforeLoan GracePeriodStart GracePeriodEnd  \
0                                    0.0       2019-10-28     2020-01-27   
1                                    0.0              NaN            NaN   
2                                    1.0              NaN            NaN   
3                                    0.0              NaN            NaN   
4                                    1.0              NaN            NaN   

  NextPaymentDate NextPaymentNr NrOfScheduledPayments  ReScheduledOn  \
0             NaN           NaN                   NaN            NaN   
1      2022-07-11           2.0                  60.0            NaN   
2             NaN           NaN                   NaN            NaN   
3             NaN           NaN                   NaN            NaN   
4             NaN           NaN                   NaN            NaN   

   PrincipalDebtServicingCost  InterestAndPenaltyDebtServicingCost  \
0                        0.00                                63.11   
1                         NaN                                  NaN   
2                        0.00                               136.62   
3                      259.67                              1659.60   
4                        0.00                               215.05   

   ActiveLateLastPaymentCategory  
0                          16-30  
1                            NaN  
2                          16-30  
3                           180+  
4                           180+  

[5 rows x 112 columns]

In [3]:
df.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 240440 entries, 0 to 240439
Data columns (total 112 columns):
 #    Column                                  Non-Null Count   Dtype  
---   ------                                  --------------   -----  
 0    ReportAsOfEOD                           240440 non-null  object 
 1    LoanId                                  240440 non-null  object 
 2    LoanNumber                              240440 non-null  int64  
 3    ListedOnUTC                             240440 non-null  object 
 4    BiddingStartedOn                        240440 non-null  object 
 5    BidsPortfolioManager                    240440 non-null  int64  
 6    BidsApi                                 240440 non-null  float64
 7    BidsManual                              240440 non-null  float64
 8    PartyId                                 240440 non-null  object 
 9    NewCreditCustomer                       240440 non-null  bool   
 10   LoanApplicationStartedDate    

In [4]:
pd.options.display.max_rows = 4000
unique = df.nunique()
unique

ReportAsOfEOD                                  1
LoanId                                    240440
LoanNumber                                240440
ListedOnUTC                               239171
BiddingStartedOn                          239172
BidsPortfolioManager                        5814
BidsApi                                     1287
BidsManual                                  4461
PartyId                                   111401
NewCreditCustomer                              2
LoanApplicationStartedDate                239304
LoanDate                                    4367
ContractEndDate                             4670
FirstPaymentDate                            2691
MaturityDate_Original                       3859
MaturityDate_Last                           4630
ApplicationSignedHour                         24
ApplicationSignedWeekday                       7
VerificationType                               5
LanguageCode                                  13
Age                 

selecting the relevant columns after reading the dataset description

In [5]:
df = df[['Age', 'Gender', 'VerificationType','NewCreditCustomer', 'AppliedAmount', 
'UseOfLoan','MonthlyPayment', 'Country', 'Interest','LoanDuration', 'Education', 'MaritalStatus', 'NrOfDependants', 
'EmploymentStatus', 'EmploymentDurationCurrentEmployer', 'WorkExperience', 'OccupationArea', 
'HomeOwnershipType', 'IncomeTotal', 'ExistingLiabilities', 'LiabilitiesTotal', 'RefinanceLiabilities', 
'DebtToIncome', 'FreeCash', 'CreditScoreEsMicroL', 'CreditScoreEsEquifaxRisk', 
'CreditScoreFiAsiakasTietoRiskGrade', 'CreditScoreEeMini', 'NoOfPreviousLoansBeforeLoan', 
'AmountOfPreviousLoansBeforeLoan', 'PreviousRepaymentsBeforeLoan', 'PreviousEarlyRepaymentsBefoleLoan',
'PreviousEarlyRepaymentsCountBeforeLoan', 'DefaultDate', 'Status']]

In [6]:
df.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 240440 entries, 0 to 240439
Data columns (total 35 columns):
 #   Column                                  Non-Null Count   Dtype  
---  ------                                  --------------   -----  
 0   Age                                     240440 non-null  int64  
 1   Gender                                  240395 non-null  float64
 2   VerificationType                        240390 non-null  float64
 3   NewCreditCustomer                       240440 non-null  bool   
 4   AppliedAmount                           240440 non-null  float64
 5   UseOfLoan                               240440 non-null  int64  
 6   MonthlyPayment                          233750 non-null  float64
 7   Country                                 240440 non-null  object 
 8   Interest                                240440 non-null  float64
 9   LoanDuration                            240440 non-null  int64  
 10  Education                               2403

checking the number of missing values

In [7]:
df.isnull().sum()

Age                                            0
Gender                                        45
VerificationType                              50
NewCreditCustomer                              0
AppliedAmount                                  0
UseOfLoan                                      0
MonthlyPayment                              6690
Country                                        0
Interest                                       0
LoanDuration                                   0
Education                                     50
MaritalStatus                                 50
NrOfDependants                            204840
EmploymentStatus                             202
EmploymentDurationCurrentEmployer           3358
WorkExperience                            203918
OccupationArea                                91
HomeOwnershipType                           1657
IncomeTotal                                    0
ExistingLiabilities                            0
LiabilitiesTotal    

Exploring the status column

In [8]:
df['Status'].unique()

array(['Late', 'Current', 'Repaid'], dtype=object)

In [9]:
df.isnull().sum()

Age                                            0
Gender                                        45
VerificationType                              50
NewCreditCustomer                              0
AppliedAmount                                  0
UseOfLoan                                      0
MonthlyPayment                              6690
Country                                        0
Interest                                       0
LoanDuration                                   0
Education                                     50
MaritalStatus                                 50
NrOfDependants                            204840
EmploymentStatus                             202
EmploymentDurationCurrentEmployer           3358
WorkExperience                            203918
OccupationArea                                91
HomeOwnershipType                           1657
IncomeTotal                                    0
ExistingLiabilities                            0
LiabilitiesTotal    

Get the percentage of missing values for each column 

In [10]:
pd.set_option('display.max_rows', None)
df.isnull().sum()/len(df)* 100

Age                                        0.000000
Gender                                     0.018716
VerificationType                           0.020795
NewCreditCustomer                          0.000000
AppliedAmount                              0.000000
UseOfLoan                                  0.000000
MonthlyPayment                             2.782399
Country                                    0.000000
Interest                                   0.000000
LoanDuration                               0.000000
Education                                  0.020795
MaritalStatus                              0.020795
NrOfDependants                            85.193811
EmploymentStatus                           0.084013
EmploymentDurationCurrentEmployer          1.396606
WorkExperience                            84.810348
OccupationArea                             0.037847
HomeOwnershipType                          0.689153
IncomeTotal                                0.000000
ExistingLiab

### Dropping columns with lots of missing values. 

For loans that are late, they cannot be classified as defaulted because although the repayment is just late and not defaulted. A loan becomes defaulted when the borrower fails to make payment 60 days after the grace period (according to https://ccr.equifax.com.au/resources/explaining-the-difference-between-a-late-payment-and-a-default-to-your-customers#:~:text=Repayment%20history%20information%20is%20recorded,was%20in%20a%20specific%20month.&text=A%20payment%20later%20than%2060%20days%20is%20a%20default.) Since this date can only be defined by the loan company, this might be difficult to achieve. However, the loan company has provided us with a Defaultdate column which is the date the loan went into defaulted state. 

This means for rows or borrowers that have a default date, the borrower is assumed to have defaulted the payment of the loan. This means that the default date is what will be used to determine whether a borrower has defaulted a loan or not. This also means that the defaultdate column cannot be dropped. 

In [11]:
df = df.drop(columns=['NrOfDependants', 'WorkExperience', 'CreditScoreEsMicroL', 
'CreditScoreEsEquifaxRisk', 'CreditScoreFiAsiakasTietoRiskGrade', 'CreditScoreEeMini', 
'PreviousRepaymentsBeforeLoan', 'PreviousEarlyRepaymentsBefoleLoan'])
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 240440 entries, 0 to 240439
Data columns (total 27 columns):
 #   Column                                  Non-Null Count   Dtype  
---  ------                                  --------------   -----  
 0   Age                                     240440 non-null  int64  
 1   Gender                                  240395 non-null  float64
 2   VerificationType                        240390 non-null  float64
 3   NewCreditCustomer                       240440 non-null  bool   
 4   AppliedAmount                           240440 non-null  float64
 5   UseOfLoan                               240440 non-null  int64  
 6   MonthlyPayment                          233750 non-null  float64
 7   Country                                 240440 non-null  object 
 8   Interest                                240440 non-null  float64
 9   LoanDuration                            240440 non-null  int64  
 10  Education                               2403

Creating the default column using the rows that have a defaultdate value

In [12]:
df['DefaultDate'].isnull()


df['Default'] = df['DefaultDate'].fillna('').apply(lambda x: 0 if x=='' else 1)
df['Default'].value_counts()

0    159123
1     81317
Name: Default, dtype: int64

This means 150123 borrowers did not default on their loan since their default date is null. This is also confirmed below.

0 = Did not default
1 = Defaulted

In [13]:
df['DefaultDate'].isnull().sum()

159123

In [14]:
df.corr()

Age    Gender  VerificationType  \
Age                                     1.000000  0.033175         -0.030722   
Gender                                  0.033175  1.000000          0.125324   
VerificationType                       -0.030722  0.125324          1.000000   
NewCreditCustomer                      -0.049172  0.111295          0.096072   
AppliedAmount                           0.089733 -0.027141          0.038192   
UseOfLoan                              -0.092404  0.040499         -0.151672   
MonthlyPayment                          0.048114  0.008426         -0.042793   
Interest                               -0.046524  0.171666          0.005632   
LoanDuration                            0.066457  0.065948          0.143126   
Education                               0.083127  0.069278         -0.005068   
MaritalStatus                          -0.079040  0.042586         -0.145770   
EmploymentStatus                       -0.055234  0.035438         -0.158114   
OccupationArea                         -0.056653  0.050768         -0.127936   
HomeOwnershipType                      -0.029179  0.022409          0.022845   
IncomeTotal                             0.018240 -0.013153          0.010687   
ExistingLiabilities                     0.069228 -0.052656         -0.081974   
LiabilitiesTotal                        0.003467  0.004323          0.000613   
RefinanceLiabilities                   -0.029197  0.034473          0.039977   
DebtToIncome                           -0.056444  0.066555         -0.080005   
FreeCash                               -0.013912 -0.005442         -0.068783   
NoOfPreviousLoansBeforeLoan             0.045538 -0.080302         -0.120113   
AmountOfPreviousLoansBeforeLoan         0.099155 -0.097857         -0.082252   
PreviousEarlyRepaymentsCountBeforeLoan -0.000702 -0.030940         -0.053974   
Default                                 0.007479  0.032530         -0.138522   

                                        NewCreditCustomer  AppliedAmount  \
Age                                             -0.049172       0.089733   
Gender                                           0.111295      -0.027141   
VerificationType                                 0.096072       0.038192   
NewCreditCustomer                                1.000000       0.088616   
AppliedAmount                                    0.088616       1.000000   
UseOfLoan                                        0.065061      -0.006766   
MonthlyPayment                                   0.142166       0.789767   
Interest                                         0.226654      -0.030791   
LoanDuration                                    -0.033420       0.255787   
Education                                       -0.069158       0.035395   
MaritalStatus                                    0.101175       0.029690   
EmploymentStatus                                 0.099964       0.043466   
OccupationArea                                   0.074373       0.034921   
HomeOwnershipType                                0.025874       0.015089   
IncomeTotal                                      0.011514       0.016015   
ExistingLiabilities                             -0.386865      -0.084835   
LiabilitiesTotal                                -0.000645      -0.000917   
RefinanceLiabilities                             0.054519       0.090548   
DebtToIncome                                     0.049394       0.114961   
FreeCash                                         0.041497       0.033975   
NoOfPreviousLoansBeforeLoan                     -0.603651      -0.157201   
AmountOfPreviousLoansBeforeLoan                 -0.625608      -0.074147   
PreviousEarlyRepaymentsCountBeforeLoan          -0.104904      -0.004650   
Default                                          0.112726       0.071169   

                                        UseOfLoan  MonthlyPayment  Interest  \
Age                                     -0.092404        0.04811

Since the status column for a new applicant cannot be known, it will be dropped. 

Also, since the default date cannot be known and it has been used to determine if the applicant defaulted or not, it will also be dropped. 

We can also drop the country column as it doesn't really bears no relevance to determining if an individual will default on his/her loan repayment. 

In [15]:
df.drop(columns=['Status', 'DefaultDate'], inplace=True)
len(df.columns)

26

In [16]:
df.head()

Age  Gender  VerificationType  NewCreditCustomer  AppliedAmount  UseOfLoan  \
0   53     1.0               4.0              False         2125.0          2   
1   25     0.0               4.0              False          531.0         -1   
2   50     1.0               1.0              False         3000.0          3   
3   44     0.0               4.0               True        10630.0          3   
4   42     0.0               3.0               True         1500.0          2   

   MonthlyPayment Country  Interest  LoanDuration  ...  IncomeTotal  \
0           62.05      EE     20.97            60  ...        354.0   
1           16.67      EE     25.27            60  ...        947.0   
2           84.75      EE     17.12            60  ...        900.0   
3          268.57      EE     13.67            60  ...       1200.0   
4           63.53      ES     40.40            60  ...        863.0   

   ExistingLiabilities  LiabilitiesTotal RefinanceLiabilities  DebtToIncome  \
0                    8            485.09                    6         26.29   
1                    0              0.00                    0          0.00   
2                    4            736.45                    0         30.58   
3                    7            905.00                    3         26.71   
4                    1            350.00                    0          7.36   

   FreeCash  NoOfPreviousLoansBeforeLoan  AmountOfPreviousLoansBeforeLoan  \
0     10.92                          1.0                            500.0   
1      0.00                          8.0                           8609.0   
2     78.80                          1.0                           1800.0   
3    349.43                          0.0                              0.0   
4    449.47                          0.0                              0.0   

   PreviousEarlyRepaymentsCountBeforeLoan  Default  
0                                     0.0        1  
1                                     0.0        0  
2                                     1.0        1  
3                                     0.0        1  
4                                     1.0        1  

[5 rows x 26 columns]

In [19]:
len(df)

240440

In [20]:
df.isna().sum()


Age                                          0
Gender                                      45
VerificationType                            50
NewCreditCustomer                            0
AppliedAmount                                0
UseOfLoan                                    0
MonthlyPayment                            6690
Country                                      0
Interest                                     0
LoanDuration                                 0
Education                                   50
MaritalStatus                               50
EmploymentStatus                           202
EmploymentDurationCurrentEmployer         3358
OccupationArea                              91
HomeOwnershipType                         1657
IncomeTotal                                  0
ExistingLiabilities                          0
LiabilitiesTotal                             0
RefinanceLiabilities                         0
DebtToIncome                                50
FreeCash     

In [24]:
df["EmploymentDurationCurrentEmployer"].head()

0    MoreThan5Years
1        UpTo5Years
2    MoreThan5Years
3        UpTo3Years
4        UpTo5Years
Name: EmploymentDurationCurrentEmployer, dtype: object

- Looking at the data, I will be dropping columns that the monthly payment is unknown, because that will be difficult to infer as this is heavily dependent on a lot of factors (e.g. amount the person is borrowing, what he wants it for, his salary etc.)

- For the null values in gender, I will assign all the null values to unknown which is 2 from the dataset. 
- For verification type, all null values will become not set which is 0. 
- For education, I will replace all the educations with the modal (highest occurence value) value in that column. 
- I will also replace the missing marital status values with the modal value of that column 
- I will replace the missing employment status values with the modal values. 
- I will replace the missing EmploymentDurationCurrentEmployer values with the modal values in that column. 
- I will replace missing occupationArea with other which is 0. 
- I will replace missing Homeownership type with other which is 10. 
- I will replace missing debtToIncome values with the mean value of that column. 
- For the NoOfPreviousLoansBeforeLoan, AmountOfPreviousLoansBeforeLoan and PreviousEarlyRepaymentsCountBeforeLoan I will replace their values with the modal values in their column. 

#### This way, I will be losing about 6690 rows of the data, only from the monthly payment column. Some of the other columns may not need to be dropped as some of the data may be in the same row that is being dropped once the rows with empty datasets are dropped from the monthly payment column.  


In [22]:
# Dropping null values in Monthly payment.

df.dropna(subset=['MonthlyPayment'], inplace=True)

In [23]:
df.isna().sum()

Age                                          0
Gender                                       0
VerificationType                             0
NewCreditCustomer                            0
AppliedAmount                                0
UseOfLoan                                    0
MonthlyPayment                               0
Country                                      0
Interest                                     0
LoanDuration                                 0
Education                                    0
MaritalStatus                                0
EmploymentStatus                             0
EmploymentDurationCurrentEmployer         2648
OccupationArea                               0
HomeOwnershipType                            0
IncomeTotal                                  0
ExistingLiabilities                          0
LiabilitiesTotal                             0
RefinanceLiabilities                         0
DebtToIncome                                 0
FreeCash     

Almost all the null values from the other columns have been removed just by dropping the null values in the monthly payment column. I will therefore implement my initial plans on the remaining null values. 


In [27]:
df['EmploymentDurationCurrentEmployer'].fillna(df['EmploymentDurationCurrentEmployer'].mode()[0], inplace=True)
df['NoOfPreviousLoansBeforeLoan'].fillna(df['NoOfPreviousLoansBeforeLoan'].mode()[0], inplace=True)
df['AmountOfPreviousLoansBeforeLoan'].fillna(df['AmountOfPreviousLoansBeforeLoan'].mode()[0], inplace=True)
df['PreviousEarlyRepaymentsCountBeforeLoan'].fillna(df['PreviousEarlyRepaymentsCountBeforeLoan'].mode()[0], inplace=True)

In [28]:
df.isna().sum()

Age                                       0
Gender                                    0
VerificationType                          0
NewCreditCustomer                         0
AppliedAmount                             0
UseOfLoan                                 0
MonthlyPayment                            0
Country                                   0
Interest                                  0
LoanDuration                              0
Education                                 0
MaritalStatus                             0
EmploymentStatus                          0
EmploymentDurationCurrentEmployer         0
OccupationArea                            0
HomeOwnershipType                         0
IncomeTotal                               0
ExistingLiabilities                       0
LiabilitiesTotal                          0
RefinanceLiabilities                      0
DebtToIncome                              0
FreeCash                                  0
NoOfPreviousLoansBeforeLoan     

### Categorical vs Continous variables
I will need to convert all the categoriacl variables to features. Looking at the datasets, data like gender, verificationtype, newcreditcustomer etc, are categorical variables as they have meanings beyond being just numbers. Income total however is a value and not a category of values. 

I will be using one hot encoding instead of label encoding to convert all the categorical variables into numerical variable. Although label encoding is more direct and easy to implement, it can throw off the model as it assigns values to numbers based on their alphabetical order, the model can therefore assume that 4 is greater than 1 and make a prediction based on that.

One hot encoding however, creates additional features based on the number of unique values in the categorical feature. It has the disadvantage of adding more columns to the dataset. Every unique value in the category will be added as a feature.

Identifying the categorical variables.

The categorical variables are: 

<li>Gender</li>
<li>VerificationType </li>
<li>NewCreditCustomer</li>
<li>UseOfLoan</li>
<li>Country </li>
<li>LoanDuration </li>
<li>Education</li>
<li>MaritalStatus</li>
<li>EmploymentStatus</li>
<li>EmploymentDurationCurrentEmployer</li>
<li>OccupationArea</li>
<li>HomeOwnershipType</li>

The default column will be left alone since it only involves two variables 1 and 0. Similar to what's expected from label encoding. Using label encoding this way is fine. 

In [38]:
from sklearn.preprocessing import OneHotEncoder
enc=OneHotEncoder()

df = df.reset_index(drop=True)
#Passing encoded columns
col_names = ['Gender', 'VerificationType', 'NewCreditCustomer', 'UseOfLoan', 'Country', 
'LoanDuration', 'Education', 'MaritalStatus', 'EmploymentStatus', 'OccupationArea', 'HomeOwnershipType', 'EmploymentDurationCurrentEmployer']

enc_data =pd.DataFrame(enc.fit_transform(df[['Gender', 'VerificationType', 'NewCreditCustomer', 'UseOfLoan', 'Country', 
'LoanDuration', 'Education', 'MaritalStatus', 'EmploymentStatus', 'OccupationArea', 'HomeOwnershipType', 'EmploymentDurationCurrentEmployer']]).toarray(), columns=enc.get_feature_names(col_names))

#Merge with main
res_df = df.join(enc_data)
res_df.drop(columns= ['Gender', 'VerificationType', 'NewCreditCustomer', 'UseOfLoan', 'Country', 
'LoanDuration', 'Education', 'MaritalStatus', 'EmploymentStatus', 'OccupationArea', 
'HomeOwnershipType', 'EmploymentDurationCurrentEmployer'], inplace = True)
res_df.head()

c:\Users\EliteBook\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\utils\deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


Age  AppliedAmount  MonthlyPayment  Interest  IncomeTotal  \
0   53         2125.0           62.05     20.97        354.0   
1   25          531.0           16.67     25.27        947.0   
2   50         3000.0           84.75     17.12        900.0   
3   44        10630.0          268.57     13.67       1200.0   
4   42         1500.0           63.53     40.40        863.0   

   ExistingLiabilities  LiabilitiesTotal  RefinanceLiabilities  DebtToIncome  \
0                    8            485.09                     6         26.29   
1                    0              0.00                     0          0.00   
2                    4            736.45                     0         30.58   
3                    7            905.00                     3         26.71   
4                    1            350.00                     0          7.36   

   FreeCash  ...  HomeOwnershipType_10.0  \
0     10.92  ...                     0.0   
1      0.00  ...                     0.0   
2     78.80  ...                     0.0   
3    349.43  ...                     0.0   
4    449.47  ...                     0.0   

   EmploymentDurationCurrentEmployer_MoreThan5Years  \
0                                               1.0   
1                                               0.0   
2                                               1.0   
3                                               0.0   
4                                               0.0   

   EmploymentDurationCurrentEmployer_Other  \
0                                      0.0   
1                                      0.0   
2                                      0.0   
3                                      0.0   
4                                      0.0   

   EmploymentDurationCurrentEmployer_Retiree  \
0                                        0.0   
1                                        0.0   
2                                        0.0   
3                                        0.0   
4                                        0.0   

   EmploymentDurationCurrentEmployer_TrialPeriod  \
0                                            0.0   
1                                            0.0   
2                                            0.0   
3                                            0.0   
4                                            0.0   

   EmploymentDurationCurrentEmployer_UpTo1Year  \
0                                          0.0   
1                                          0.0   
2                                          0.0   
3                                          0.0   
4                                          0.0   

   EmploymentDurationCurrentEmployer_UpTo2Years  \
0                                           0.0   
1                                           0.0   
2                                           0.0   
3                                           0.0   
4                                           0.0   

   EmploymentDurationCurrentEmployer_UpTo3Years  \
0                                           0.0   
1                                           0.0   
2                                           0.0   
3                                           1.0   
4                                           0.0   

   EmploymentDurationCurrentEmployer_UpTo4Years  \
0                                           0.0   
1                                           0.0   
2                                           0.0   
3                                           0.0   
4                                           0.0   

   EmploymentDurationCurrentEmployer_UpTo5Years  
0                                           0.0  
1                                           1.0  
2                                           0.0  
3                                           0.0  
4                                           1.0  

[5 rows x 140 columns]

Splitting the entire dataset to test and train data.

In [39]:
# Extracting the dependent and independent variables. ALl the variables are independent except the default column which we derived earlier.

y = res_df['Default']
X = res_df.loc[:, res_df.columns != 'Default']

In [40]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(175312, 139) (58438, 139) (175312,) (58438,)


## Random Forest

In [41]:
# Building the model

from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import mean_absolute_error


model = RandomForestClassifier(n_estimators=100, max_depth=10, random_state=0).fit(X_train, y_train)
predictions = model.predict(X_test)
accuracy = accuracy_score(predictions, y_test)
print('Accuracy Score is', accuracy)

rf_val_mae = mean_absolute_error(predictions, y_test)

print("Validation MAE for Random Forest Model: {}".format(rf_val_mae))

Accuracy Score is 0.7655635031999726
Validation MAE for Random Forest Model: 0.23443649680002737


#### Getting the best n_estimator to use for the randomforest model

In [42]:
def get_mae(estimator, X_train, X_test, y_train, y_test):
    model = RandomForestClassifier(n_estimators=estimator, max_depth=10, random_state=0).fit(X_train, y_train)
    predictions = model.predict(X_test)
    mae = mean_absolute_error(predictions, y_test)
    return(mae)

In [43]:
candidate_estimators = [100, 200, 300, 400, 500, 1000]

values = {estimator :  get_mae(estimator, X_train, X_test, y_train, y_test)
          for estimator in candidate_estimators}

# print(values)

best_estimator = min(values, key = values.get)
print("The best n_estimator values is", best_estimator)

{100: 0.23443649680002737, 200: 0.23450494541223177, 300: 0.2343167117286697, 400: 0.23411136589205653, 500: 0.2343851603408741, 1000: 0.23464184263664054}
400


#### Getting the best depth to use for the random forest model

In [44]:
def get_mae(depth, X_train, X_test, y_train, y_test):
    model = RandomForestClassifier(n_estimators=100, max_depth=depth, random_state=0).fit(X_train, y_train)
    predictions = model.predict(X_test)
    mae = mean_absolute_error(predictions, y_test)
    return(mae)

In [45]:
candidate_depths = [1, 2, 4, 6, 8, 10, 12, 14, 20, 50, 100]
right_depth = {depth: get_mae(depth, X_train, X_test, y_train, y_test) for depth in candidate_depths}
best_depth = min(right_depth, key = right_depth.get)
print("The best depth is", best_depth)

The best depth is 100


## Final Model selction for random forest.

Based on the provided parameters above, I will select the parameters that are regarded as the best parameters to use and then use them on the actual model below. 

In [46]:
model = RandomForestClassifier(n_estimators=400, max_depth=100, random_state=0).fit(X_train, y_train)
predictions = model.predict(X_test)
accuracy = accuracy_score(predictions, y_test)
print('Accuracy Score is', accuracy)

rf_val_mae = mean_absolute_error(predictions, y_test)

print("Validation MAE for Random Forest Model: {}".format(rf_val_mae))

Accuracy Score is 0.7915568636845888
Validation MAE for Random Forest Model: 0.2084431363154112


### Decision Tree
Guide for selecting parameters: https://towardsdatascience.com/how-to-tune-a-decision-tree-f03721801680

I will be using the same random state that I used in random forest in order for me to work with the same data randomness in both models

In [47]:
from sklearn.tree import DecisionTreeClassifier
clf = DecisionTreeClassifier(random_state = 0).fit(X_train, y_train)
predictor = clf.predict(X_test)
accuracy = accuracy_score(predictor, y_test)
print('Accuracy Score is', accuracy)


rf_val_mae = mean_absolute_error(predictor, y_test)

print("Validation MAE for Decision Tree Model: {}".format(rf_val_mae))

Accuracy Score is 0.7176665868099524
Validation MAE for Decision Tree Model: 0.28233341319004757


In [48]:
def get_mae_dtc(depth, X_train, X_test, y_train, y_test):
    clf = DecisionTreeClassifier(max_depth = depth, random_state = 0).fit(X_train, y_train)
    predictor = clf.predict(X_test)
    rf_val_mae = mean_absolute_error(predictor, y_test)
    return(rf_val_mae)

In [49]:
candidate_depths = [1, 2, 4, 6, 8, 10, 12, 13, 14, 20, 50, 100]
right_depth = {depth: get_mae_dtc(depth, X_train, X_test, y_train, y_test) for depth in candidate_depths}
print(right_depth)
best_depth = min(right_depth, key = right_depth.get)
print("The best depth is", best_depth)

{1: 0.3387008453403607, 2: 0.244789349395941, 4: 0.2370717683698963, 6: 0.2331530853211951, 8: 0.2277627571100996, 10: 0.2282761217016325, 12: 0.2309627297306547, 13: 0.23296485163763306, 14: 0.23688353468633425, 20: 0.2557753516547452, 50: 0.28296656285293814, 100: 0.28233341319004757}
The best depth is 8


## Final Model selction for Decision Tree.

Based on the provided parameters above, I will select the parameters that are regarded as the best parameters to use and then use them on the actual model below. 

In [50]:
from sklearn.tree import DecisionTreeClassifier
clf = DecisionTreeClassifier(max_depth = 8, random_state = 0).fit(X_train, y_train)
predictor = clf.predict(X_test)
accuracy = accuracy_score(predictor, y_test)
print('Accuracy Score is', accuracy)


rf_val_mae = mean_absolute_error(predictor, y_test)

print("Validation MAE for K Decision Tree Model: {}".format(rf_val_mae))

Accuracy Score is 0.7722372428899004
Validation MAE for K Decision Tree Model: 0.2277627571100996


### KNN

In [51]:
from sklearn.neighbors import KNeighborsClassifier
model = KNeighborsClassifier(n_neighbors=200).fit(X_train, y_train)
predictor = model.predict(X_test)
accuracy = accuracy_score(predictor, y_test)
print('Accuracy Score is', accuracy)

Accuracy Score is 0.714586399260755


In [55]:
def get_mae_dtc(neighbor, X_train, X_test, y_train, y_test):
    clf = KNeighborsClassifier(n_neighbors=neighbor).fit(X_train, y_train)
    predictor = clf.predict(X_test)
    rf_val_mae = mean_absolute_error(predictor, y_test)
    return(rf_val_mae)

In [56]:
candidate_neighbors = np.arange(1, 500, 10)
right_neighbor = {depth: get_mae_dtc(depth, X_train, X_test, y_train, y_test) for depth in candidate_neighbors}
print(right_neighbor)
best_neighbor = min(right_neighbor, key = right_neighbor.get)
print("the best neighbor value is", best_neighbor)

{1: 0.33919709777884255, 11: 0.28127245970087955, 21: 0.27518053321468905, 31: 0.27408535541941886, 41: 0.27348643006263046, 51: 0.2738457852767035, 61: 0.27531743043909784, 71: 0.2761901502447038, 81: 0.2767206269892878, 91: 0.2772511037338718, 101: 0.277644683254047, 111: 0.2784318422943975, 121: 0.2793730107122078, 131: 0.28029706697696705, 141: 0.2811868989356241, 151: 0.2817173756802081, 161: 0.28247031041445636, 171: 0.2830863479242958, 181: 0.28347992744447104, 191: 0.28454088093363905, 201: 0.28539648858619393, 211: 0.28632054485095315, 221: 0.28688524590163933, 231: 0.2872103768096102, 241: 0.28750128341147885, 251: 0.28794619939080734, 261: 0.2881344330743694, 271: 0.28801464800301174, 281: 0.2889558164208221, 291: 0.28926383517574183, 301: 0.289298059481844, 311: 0.28946918101235497, 321: 0.289828536226428, 331: 0.2896231903898148, 341: 0.2894007324001506, 351: 0.2895547417776105, 361: 0.2894862931654061, 371: 0.29003388206304115, 381: 0.289828536226428, 391: 0.2902050035935

## Final Model selction for KNN.

Based on the provided parameters above, I will select the parameters that are regarded as the best parameters to use and then use them on the actual model below. 

In [57]:
from sklearn.neighbors import KNeighborsClassifier
model = KNeighborsClassifier(n_neighbors=41).fit(X_train, y_train)
predictor = model.predict(X_test)
accuracy = accuracy_score(predictor, y_test)
print('Accuracy Score is', accuracy)

rf_val_mae = mean_absolute_error(predictor, y_test)

print("Validation MAE for K Nearest Neighbor Model: {}".format(rf_val_mae))

Accuracy Score is 0.7265135699373695
Validation MAE for K Nearest Neighbor Model: 0.27348643006263046
